extensions/functions/src/test/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregationIntegrationTest.java

```diff
@@ -0,0 +1,51 @@
+package io.crate.operation.aggregation;
+
+import io.crate.integrationtests.SQLIntegrationTestCase;
+import io.crate.testing.TestingHelpers;
+import io.crate.window.EnterpriseFunctionsProxyTestPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import static org.hamcrest.CoreMatchers.is;
+
+public class HyperLogLogDistinctAggregationIntegrationTest extends SQLIntegrationTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        ArrayList<Class<? extends Plugin>> plugins = new ArrayList<>(super.nodePlugins());
+        plugins.add(EnterpriseFunctionsProxyTestPlugin.class);
+        return plugins;
+    }
+
+    @Test
+    public void test_hyperloglog_distinct_string() {
+        execute("create table tbl (x text)");
+        int bulkSize = 10000;
+        Object[][] bulkArgs = new Object[bulkSize][];
+        for (int i = 0; i < bulkSize; i++) {
+            bulkArgs[i] = new Object[]{Integer.toString(i)};
+            // For simplicity of actual distinct computation all elements are distinct.
+            // If data generation gets changed than error computation below has to be updated.
+            // Currently actual number of distinct is equal to bulkSize.
+        }
```

Comment:
What's the reason for the large bulk size? This test will be rather expensive.

Aren't the unit-tests enough to cover the new optimizations?

This seems to be partially a new functional/behavior test. Not sure if that is needed as long as we assert that optimized execution yields the same results as the previous executions.

Reply:
Did you investigate why it is `null` there? We should ensure that the aggregation test framework tests the optimized execution paths.

Reply:
It seems like inserted String values are not flushed. I tried to throw an Exception on `partialResultWithDocValues == null && implementation != null` and found out that [Count(String)](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/test/java/io/crate/execution/engine/aggregation/impl/CountAggregationTest.java#L104) and [Arbitrary(String)](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/test/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregationTest.java#L110) are also failing with that check. DocValues [are empty](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java#L177) but only for `String` type.

Reply:
The problem isn't that the records are not inserted, but that it doesn't use the optimized path because there are no doc-values available. 

The reason for that is that we create the mapping definition as `text`:

https://github.com/crate/crate/blob/a40419673ac591c22d72252cea1d769b611699c3/server/src/testFixtures/java/io/crate/operation/aggregation/AggregationTestCase.java#L393-L394

This should be `keyword` for strings that are not analyzed.

Reply:
it worked, thanks!